CHAT_SERVER/chat_server.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"bufio"
+	"net"
+	"strings"
+)
+
+const (
+	CONN_TYPE = "tcp"
+)
+
+type Client struct {
+	id       string
+	name     string
+	incoming chan string
+	reader   *bufio.Reader
+	writer   *bufio.Writer
+}
+
+type Room struct {
+	clients   map[string]*Client
+	name      string
+	connJoins chan net.Conn
+	incoming  chan string
+}
+
+func (room *Room) post(message string) {
+	for _, client := range room.clients {
+		index := strings.Index(message, "|")
+		if client.id != message[:index] {
+			client.writer.WriteString(message[index+1:])
+			client.writer.Flush()
+		}
+	}
+
+}
+
+func (room *Room) introducing(message string) {
+	for _, client := range room.clients {
+		client.writer.WriteString(message)
+		client.writer.Flush()
+	}
+}
+
+func (room *Room) startRoomChat() {
+	go func() {
+		for {
+			select {
+			case message := <-room.incoming:
+				room.post(message)
+			case conn := <-room.connJoins:
+				room.join(conn)
+			}
+		}
+	}()
+}
+func newClient(conn net.Conn) *Client {
+	writer := bufio.NewWriter(conn)
+	reader := bufio.NewReader(conn)
+	name, _ := reader.ReadString('\n')
+	identifier := conn.RemoteAddr().String()
+	client := &Client{
+		name:     name[:len(name)-2], //triming \n chars
+		id:       identifier,
+		incoming: make(chan string),
+		writer:   writer,
+		reader:   reader}
+
+	return client
+}
+
+func (client *Client) listen() {
```

Comment:
Call `go client.listen()` will be more efficient than initialization of anon func in `client.listen()` =)